src/Microsoft.ML.Transforms/SvmLight/SvmLightLoader.cs

```diff
@@ -0,0 +1,819 @@
+    /// indicates that the label is unknown and should be classified using semi-supervised
+    /// techniques.
+    ///
+    /// The "0 label" semi-supervised capability was rarely used and none of our learners
+    /// currently do anything like this, though it is possible we may introduce semi-supervised
+    /// learners in the future. For now this loader just parses this as a single floating point
+    /// value, period. (Which means, for our binary classifier learners, that 0 and -1 will both
+    /// be treated identically.) If we were to support this, the natural thing would be to have
+    /// an option to map 0 to NA, somehow. But practically, variants of the SVM-light format have
+    /// promulgated to the point where nearly every time 0 is used, it actually refers to a
+    /// negative example, so we may continue to accept this corruption as "correct" by default.
+    ///
+    /// The actual feature vector is specified through a series of key/value pairs. SVM-light
+    /// requires that the keys be positive, increasing integers, except for three special keys:
+    /// cost (we interpret as Weight), qid (we interpret as GroupId) and sid (we ignore these,
+    /// but might present them as a column in the future if any of our learners implement anything
+    /// resembling slack id). The value for 'cost' is float, 'qid' is a long, and 'sid' is a long
+    /// that must be positive. If these keys are specified multiple times, the last one wins.
+    ///
+    /// SVM-light, if the tail of the value is not interpretable as a number, will ignore the tail.
+    /// E.g., "5:3.14hello" will be interpreted the same as "5:3.14". I am aware of one real dataset
+    /// that took advantage of this, and for now I do not support this type of thing.
+    ///
+    /// We do not retain the restriction on keys needing to be increasing values in our loader,
+    /// due to the way we compose our feature vectors, but it will be most efficient if this policy
+    /// is still followed. If it is followed a sort will not be required.
+    ///
+    /// This loader has the special option through the <c>xf</c> option to specify a transform,
+    /// possibly trainable, to convert the raw text of the key values into the key value. The
+    /// transform, whatever it is, must in addition to user specified options accept an argument
+    /// of the form "column=Name" to identify a column to convert. Ideally there would be some
+    /// other way to specify this other than hacking arguments. The intent of this is to allow
+    /// things like string keys, a common variant of the format, but one emphatically not allowed
+    /// by the original format.
+    /// </summary>
+    public sealed class SvmLightLoader : IDataLoader<IMultiStreamSource>
+    {
+        public enum FeatureIndices
+        {
+            ZeroBased,
+            OneBased,
+            Names
+        }
+
+        public sealed class Options
```

Comment:
xml comments on all new public API

Reply:
Do these types (FeatureIndices and Options) need to be public? I don't see them being used publicly.

---
In reply to: [356301711](https://github.com/dotnet/machinelearning/pull/4190#discussion_r356301711) [](ancestors = 356301711)

Reply:
Currently they don't need to be public. There was a discussion about whether there should be a public API that takes an `Options` parameter, or a `FeatureIndices` parameter (instead of having one called `CreateSvmLightLoader` and one called `CreateSvmLightLoaderWithFeatureNames`). Do you have a preference?

---
In reply to: [356304532](https://github.com/dotnet/machinelearning/pull/4190#discussion_r356304532) [](ancestors = 356304532,356301711)

Reply:
Do all combinations of enums and options make sense together? Or do some options only make sense for certain values of the enum?

If they all make sense together then I would expose the Options. But if there are options that don’t make sense together, then I would keep it like you have it.